core/services/log/registrations.go

```diff
@@ -0,0 +1,284 @@
+			continue
+		} else {
+			logger.Tracew("LogBroadcaster: Sending out the log",
+				"logBlockNumber", log.BlockNumber, "lowestAllowedBlockNumber", metadata.lowestAllowedBlockNumber)
+		}
+
+		// make sure that this log is not sent again on the next head by increasing the newLowestAllowedBlockNumber
+		*updates = append(*updates, listenerMetadataUpdate{
+			toUpdate:                    metadata,
+			newLowestAllowedBlockNumber: log.BlockNumber + 1,
+		})
+
+		logCopy := copyLog(log)
+		decodedLog, err := r.decoders[log.Address].ParseLog(logCopy)
+		if err != nil {
+			logger.Errorw("Could not parse contract log", "error", err)
+			continue
+		}
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			listener.HandleLog(&broadcast{
+				orm:               orm,
+				latestBlockNumber: latestBlockNumber,
+				latestBlockHash:   latestHead.Hash,
+				rawLog:            logCopy,
+				decodedLog:        decodedLog,
+				jobID:             listener.JobID(),
+				jobIDV2:           listener.JobIDV2(),
+				isV2:              listener.IsV2Job(),
+			})
+		}()
+	}
+	wg.Wait()
+}
+
+/**
+	After processing the logs in this batch, the listenerMetadata structures that we touched, are updated
+  with new information about the canonical chain and the lowestAllowedBlockNumber value (higher every time) that is used to guard against double-sends
+  Note that the updates are applied only after all the logs for the (latest height - num_confirmations) head height were sent.
+*/
+func applyListenerInfoUpdates(updates []listenerMetadataUpdate, latestHead *models.Head) {
+	for _, update := range updates {
+		if update.toUpdate.lastSeenChain == nil || latestHead.IsInChain(update.toUpdate.lastSeenChain.Hash) {
+			if update.toUpdate.lastSeenChain == nil {
+				logger.Tracew("LogBroadcaster: No chain saved for this listener yet",
+					"contractAddress", update.toUpdate.opts.Contract.Address(), "numConfirmations", update.toUpdate.opts.NumConfirmations)
+			}
+			if update.toUpdate.lowestAllowedBlockNumber < update.newLowestAllowedBlockNumber {
+				update.toUpdate.lowestAllowedBlockNumber = update.newLowestAllowedBlockNumber
+			}
+		} else {
+			logger.Debugw("LogBroadcaster: Chain reorg - resetting lowestAllowedBlockNumber",
+				"blockNumber", latestHead.Number, "blockHash", latestHead.Hash,
+				"lastSeenChainNumber", update.toUpdate.lastSeenChain.Number, "lastSeenChainHash", update.toUpdate.lastSeenChain.Hash)
+			// re-org situation: the chain was changed, so we can't use the number that tracked last unprocessed height of the previous chain
+			update.toUpdate.lowestAllowedBlockNumber = 0
+		}
+		logger.Tracew("LogBroadcaster: Setting as latest head for this listener", "blockNumber", latestHead.Number, "blockHash", latestHead.Hash)
+
+		update.toUpdate.lastSeenChain = latestHead
+	}
+}
+
+func copyLog(l types.Log) types.Log {
```

Comment:
nitpick - but this seems better defined on the log type itself

Reply:
yup, true

Reply:
oh but Log is from geth actually...